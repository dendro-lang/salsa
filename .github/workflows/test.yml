name: Test

on:
  push:
    branches:
    - master
    - staging
    - trying
  pull_request:
    paths:
    - '**.rs'
    - '**/Cargo.*'
    - '.github/workflows/**.yml'
    - 'dendro-salsa-tests/tests/compile-fail/**.stderr'
  merge_group:

jobs:
  test:
    name: Test
    strategy:
      matrix:
        rust:
        - stable
        - beta
        experimental:
        - false
        include:
        - rust: nightly
          experimental: true
    continue-on-error: ${{ matrix.experimental }}
    runs-on: ubuntu-latest
    steps:
    - name: Checkout
      uses: actions/checkout@v2
    - name: Setup Rust toolchain
      uses: actions-rs/toolchain@v1
      with:
        toolchain: ${{ matrix.rust }}
        components: rustfmt, clippy
        default: true
    - name: Format
      uses: actions-rs/cargo@v1
      with:
        command: fmt
        args: -- --check
    - name: Clippy
      uses: actions-rs/cargo@v1
      with:
        command: clippy
        args: --all-features --all-targets --all
    - name: Test
      uses: actions-rs/cargo@v1
      with:
        command: test
        args: --all-features --all-targets --all
    - name: Test docs
      uses: actions-rs/cargo@v1
      with:
        command: test
        args: --doc
    - name: Check (without default features)
      uses: actions-rs/cargo@v1
      with:
        command: check
        args: --no-default-features
