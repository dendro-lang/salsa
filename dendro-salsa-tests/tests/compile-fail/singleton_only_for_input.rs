//! Compile Singleton struct test: 
//!
//! Singleton flags are only allowed for input structs. If applied on any other Salsa struct compilation must fail

use dendro_salsa_tests::{HasLogger, Logger};

use test_log::test;

#[salsa::jar(db = Db)]
struct Jar(MyInput, MyTracked, Integers, create_tracked_structs );

trait Db: salsa::DbWithJar<Jar> + HasLogger {}

#[salsa::input(singleton)]
struct MyInput {
    field: u32,
}


#[salsa::tracked(singleton)]
struct MyTracked {
    field: u32,
}


#[salsa::tracked(singleton)]
fn create_tracked_structs(db: &dyn Db, input: MyInput) -> Vec<MyTracked> {
    (0..input.field(db))
        .map(|i| MyTracked::new(db, i))
        .collect()
}

#[salsa::accumulator(singleton)]
struct Integers(u32);


#[salsa::db(Jar)]
#[derive(Default)]
struct Database {
    storage: salsa::Storage<Self>,
    logger: Logger,
}

impl salsa::Database for Database {}

impl Db for Database {}

impl HasLogger for Database {
    fn logger(&self) -> &Logger {
        &self.logger
    }
}


fn main() {}