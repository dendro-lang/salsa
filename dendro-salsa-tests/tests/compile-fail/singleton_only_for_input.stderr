error: `singleton` option not allowed here
  --> tests/compile-fail/singleton_only_for_input.rs:20:18
   |
20 | #[salsa::tracked(singleton)]
   |                  ^^^^^^^^^

error: `singleton` option not allowed here
  --> tests/compile-fail/singleton_only_for_input.rs:26:18
   |
26 | #[salsa::tracked(singleton)]
   |                  ^^^^^^^^^

error: `singleton` option not allowed here
  --> tests/compile-fail/singleton_only_for_input.rs:33:22
   |
33 | #[salsa::accumulator(singleton)]
   |                      ^^^^^^^^^

error[E0412]: cannot find type `MyTracked` in this scope
  --> tests/compile-fail/singleton_only_for_input.rs:10:21
   |
10 | struct Jar(MyInput, MyTracked, Integers, create_tracked_structs );
   |                     ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `Integers` in this scope
  --> tests/compile-fail/singleton_only_for_input.rs:10:32
   |
10 | struct Jar(MyInput, MyTracked, Integers, create_tracked_structs );
   |                                ^^^^^^^^ not found in this scope

error[E0412]: cannot find type `create_tracked_structs` in this scope
  --> tests/compile-fail/singleton_only_for_input.rs:10:42
   |
10 | struct Jar(MyInput, MyTracked, Integers, create_tracked_structs );
   |                                          ^^^^^^^^^^^^^^^^^^^^^^ not found in this scope

warning: unused import: `test_log::test`
 --> tests/compile-fail/singleton_only_for_input.rs:7:5
  |
7 | use test_log::test;
  |     ^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
